### PR TITLE
Service

### PR DESCRIPTION
**Pull request by DarkJediDJ**
# Description
This PR provides realisation of REST API, with basic GET method

## Checklist
- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No error nor warning in the console.

## Items
- [x] Added service API with GET method
- [x] New functions to GET data from web-site
- [x] Fixed package name
- [x] Added new tests for service

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+	"1": "(",
+	"2": "[",
+	"3": "{",
+	"4": ")",
+	"5": "]",
+	"6": "}",
+}
+
+type Parentheses struct {
+	Parenthese string `json:"parentheses"`
```
Comment (gordonBusyman) [inline config.go]:
```suggestion
	Parentheses string `json:"parentheses"`
```
Or give another name

File: logic.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	"strconv"
+	"strings"
+)
+
+func RandParentheses(len string) string {
+	num, err := strconv.Atoi(len)
+	if err != nil {
+		log.Fatal(err)
+	}
+	s := ""
```
Comment (gordonBusyman) [inline logic.go]: There is another type of variable initialization in go

File: logic.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package main
```
Comment (gordonBusyman) [inline logic.go]: God, please no `logic.go`

File: logic.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		log.Fatal(err)
+	}
+	s := ""
+	for i := 0; i < num; i++ {
+		s += M[strconv.Itoa(rand.Intn(6)+1)]
+	}
+	return s
+}
+
+func GetValue(n int) string {
```
Comment (gordonBusyman) [inline logic.go]: Is it related to the functionality?  Or it is just for testing?

File: logic.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		log.Fatal(err)
+	}
+	line := string(body)
+	commas := strings.LastIndex(line, ":")
+	return line[commas+2 : len(line)-2]
+}
+
+func GetStatus(n int) int {
```
Comment (gordonBusyman) [inline logic.go]: Is it related to the functionality?  Or it is just for testing?

File: service.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+
+	"github.com/gorilla/mux"
+)
+
+func returnSingleArticle(w http.ResponseWriter, r *http.Request) {
```
Comment (gordonBusyman) [inline service.go]:
```suggestion
func generate(w http.ResponseWriter, r *http.Request) {
```

File: service.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	res, err := json.Marshal(line)
+	if err != nil {
+		log.Fatal(err)
+	}
+	w.Header().Set("Content-Type", "application/json")
+	_, err = w.Write(res)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
```
Comment (gordonBusyman) [inline service.go:44]: New line here, why doesn't linter complain?

File: logic.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import (
+	"io"
+	"log"
+	"math/rand"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+func RandParentheses(len string) string {
```
Comment (gordonBusyman) [inline logic.go]: This is Public method. Please add a comment

File: service_test.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_returnSingleArticle(t *testing.T) {
+	testServiceCases := []struct {
+		value       int
+		want        int
+		description string
+	}{
+		{5, 5, "Standard parantheses string"},
```
Comment (gordonBusyman) [inline service_test.go]:
```suggestion
		{5, 5, "Standard parentheses string"},
```

**Review by gordonBusyman — CHANGES_REQUESTED** [review]
Before continue answer to the question: What is go handler?

File: service.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+	Router *mux.Router
+}
+
+func generate(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+
+	length, ok := vars["length"]
+	if !ok {
+		log.Fatal("Parameter not found")
+	}
```
Comment (gordonBusyman) [inline service.go:22]: New line here

File: service.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+	len, err := strconv.Atoi(length)
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (gordonBusyman) [inline service.go:27]: New line here

File: service.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+// Init the API
+func (a *App) New() {
+	myRouter := mux.NewRouter().StrictSlash(false)
+	myRouter.HandleFunc("/generate/{length:[0-9]+}", generate)
+
+	a.Router = myRouter
+}
+
+// And then run it
```
Comment (gordonBusyman) [inline service.go]: COMMENT

File: service.go [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+	w.Header().Set("Content-Type", "application/json")
+
+	_, err = w.Write(res)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// Init the API
```
Comment (gordonBusyman) [inline service.go]: COMMENT

File: service_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+func TestGenerateHandler(t *testing.T) {
+	testServiceCases := []struct {
+		route string
+		want  int
+		name  string
+	}{
+		{"5", 5, "Standard parentheses string"},
+		{"1000", 1000, "Long parentheses string"},
+		{"1", 1, "Short parentheses string"},
+		{"0", 0, "Emptyy parentheses string"},
```
Comment (gordonBusyman) [inline service_test.go]: Double `y`

Comment (gordonBusyman) [reply service_test.go]: And where are tests for edge cases? Like 
`{"?", 0, "Non int char"},`

File: service_test.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+var a App
+
+func TestGenerateHandler(t *testing.T) {
+	testServiceCases := []struct {
+		route string
```
Comment (gordonBusyman) [inline service_test.go]: Is that a route? I think this is length

File: service_test.go (hunk 14 above)
Comment (gordonBusyman) [reply service_test.go]: Negative value?

File: service.go [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+	length, ok := vars["length"]
+	if !ok {
+		log.Fatal("Parameter not found")
+	}
+	len, err := strconv.Atoi(length)
+	if err != nil {
+		log.Fatal(err)
+	}
+	line := Parentheses{
+		RandParentheses(len),
```
Comment (gordonBusyman) [inline service.go]: Do you handle negative values?

**Review by gordonBusyman — CHANGES_REQUESTED** [review]

File: service_test.go [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			response := httptest.NewRecorder()
+
+			a.New()
+			a.Router.ServeHTTP(response, req)
+			assert.Equal(t, tc.status, response.Code, "Expected response code %d. Got %d\n", tc.status, response.Code)
+			if tc.status == 404 {
```
Comment (gordonBusyman) [inline service_test.go]: So if we get Status Code `500` it will proceed. Is it correct?

Comment (gordonBusyman) [reply service_test.go]: What is `404`?

**Review by gordonBusyman — CHANGES_REQUESTED** [review]

File: service_test.go [hunk 18]
```diff
@@ -39,13 +40,16 @@ func TestGenerateHandler(t *testing.T) {
 			a.New()
 			a.Router.ServeHTTP(response, req)
 			assert.Equal(t, tc.status, response.Code, "Expected response code %d. Got %d\n", tc.status, response.Code)
-			if tc.status == 404 {
+
+			if tc.status != http.StatusOK {
```
Comment (gordonBusyman) [inline service_test.go:44]: Zaebok

**Review by gordonBusyman — APPROVED** [review]
